### PR TITLE
Add regex patterns to JSON schema for `Decimal` type

### LINKED ISSUE #10867: max_digits and decimal_places do not serialize to json_schema for decimal
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Not sure if this is expected or not, but my decimal constraints do not appear to make it to the Jsonschema:

### Example Code

```Python
from pydantic import TypeAdapter, Field
from typing import Annotated
from decimal import Decimal
constraints = {'max_digits': 5, 'decimal_places': 2, 'ge': 0.0, 'le': 10.0}

TypeAdapter(Annotated[Decimal, Field(**constraints)]).json_schema()
# Out[53]: {'anyOf': [{'maximum': 10.0, 'minimum': 0.0, 'type': 'number'},  {'type': 'string'}]}
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.9.2
        pydantic-core version: 2.23.4
          pydantic-core build: profile=release pgo=false
                 install path: /home/mheiser/src/taxa-backend/.venv/lib/python3.12/site-packages/pydantic
               python version: 3.12.7 (main, Oct  8 2024, 00:20:25) [Clang 18.1.8 ]
                     platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
             related packages: fastapi-0.115.4 typing_extensions-4.12.2 pydantic-settings-2.6.1 pyright-1.1.385
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by Dima-Bulavenko**
## Change Summary

This PR adds a regular expression pattern to the JSON schema of `Decimal` fields to validate their string representation according to the `max_digits` and `decimal_places` parameters of the `Field` class.

- Updated existing tests for `Decimal` fields to align with the new validation pattern.
- Added new tests to validate the pattern across different combinations of `max_digits` and `decimal_places`.

You can run the new tests with the following command:  
```bash
uv run pytest tests/test_json_schema.py::TestDecimalPattern
```

## Related issue number
fixes #11016
fixes #11420
fixes #10867

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @Viicos

### CODE REVIEW

**Comment by Dima-Bulavenko** [comment]
please review - thanks!

**Review by Viicos — COMMENTED** [review]
Thanks for the contribution. I've checked locally and the regex seems to correctly match.

Similar to https://github.com/pydantic/pydantic/pull/11420, we'd have to also add patterns when `max_digits` xor `decimal_places` is None. Would you be interested in pursuing?

File: tests/test_json_schema.py [hunk 1]
```diff
@@ -6811,3 +6946,147 @@ def test_json_schema_arguments_v3_aliases() -> None:
         },
         'required': ['b'],
     }
+
+
+class TestDecimalPattern:
```
Comment (Viicos) [inline tests/test_json_schema.py]: Let's flatten this to use functions instead of methods.

File: pydantic/json_schema.py [hunk 2]
```diff
@@ -674,7 +674,31 @@ def decimal_schema(self, schema: core_schema.DecimalSchema) -> JsonSchemaValue:
...
+            integer_places = max_digits
+
+            if isinstance(max_digits, int) and isinstance(decimal_places, int):
+                if (diff := max_digits - decimal_places) > 0:
+                    integer_places = diff
+                else:
+                    integer_places = 0
+
+            pattern = (
+                r'^(?!^[+-\.]*$)'  # check string is not empty and not single or sequence of ".+-" characters.
```
Comment (Viicos) [inline pydantic/json_schema.py]: I think you don't need to escape the `.` when inside brackets.

File: tests/test_json_schema.py (hunk 1 above)
Comment (Dima-Bulavenko) [reply tests/test_json_schema.py]: Resolved with [this commit](https://github.com/pydantic/pydantic/pull/11987/commits/73afd017f4cb1683540cdc08f3601dc90886cd61)

**Comment by Dima-Bulavenko** [comment]
> Thanks for the contribution. I've checked locally and the regex seems to correctly match.
> 
> Similar to #11420, we'd have to also add patterns when `max_digits` xor `decimal_places` is None. Would you be interested in pursuing?

Thanks for reviewing!

It seems the pattern already handles cases where `max_digits` or `decimal_places` is `None`.

I've added tests to cover these scenarios:
```python
@pytest.fixture
def get_decimal_pattern():
    def pattern(max_digits=None, decimal_places=None) -> str:
        filed = TypeAdapter(Annotated[Decimal, Field(max_digits=max_digits, decimal_places=decimal_places)])
        return filed.json_schema()['anyOf'][1]['pattern']

    return pattern

@pytest.mark.parametrize("valid_decimal", ["0.1", "0000.1", "11.1", "0011.1", "11111111.1"])
def test_decimal_pattern_max_digits_set_none(valid_decimal, get_decimal_pattern):
    max_digits = None
    decimal_places = 1
    pattern = get_decimal_pattern(max_digits=max_digits, decimal_places=decimal_places)
    
    assert re.fullmatch(pattern, valid_decimal) is not None


@pytest.mark.parametrize("valid_decimal", ["0.1", "000.1", "0.001000", "0000.001000"])
def test_decimal_pattern_decimal_places_set_none(valid_decimal, get_decimal_pattern):
    max_digits = 3
    decimal_places = None
    pattern = get_decimal_pattern(max_digits=max_digits, decimal_places=decimal_places)
    
    assert re.fullmatch(pattern, valid_decimal) is not None

@pytest.mark.parametrize("unvalid_decimal", ["0.0001", "111.1", "11.01"])
def test_decimal_pattern_with_unvalid_decimals_decimal_places_set_none(unvalid_decimal, get_decimal_pattern):
    max_digits = 3
    decimal_places = None
    pattern = get_decimal_pattern(max_digits=max_digits, decimal_places=decimal_places)
    
    assert re.fullmatch(pattern, unvalid_decimal) is None

@pytest.mark.parametrize("valid_decimal", ["11111111", "1111.11111", "0.00000001"])
def test_decimal_pattern_decimal_places_max_digits_set_none(valid_decimal, get_decimal_pattern):
    max_digits = None
    decimal_places = None
    pattern = get_decimal_pattern(max_digits=max_digits, decimal_places=decimal_places)
    
    assert re.fullmatch(pattern, valid_decimal) is not None
```

File: pydantic/json_schema.py (hunk 2 above)
Comment (Dima-Bulavenko) [reply pydantic/json_schema.py]: Thanks for pointing that out! I’ve resolved it in commit https://github.com/pydantic/pydantic/commit/5874292e75f955fe80334c7c083be0832edcf0df.

**Comment by Viicos** [comment]
> It seems the pattern already handles cases where `max_digits` or `decimal_places` is `None`.

Sorry what I said was confusing, I wanted to say that it would be nice to have simpler patterns generated for cases where `max_digits` xor `decimal_places` is `None`.

**Comment by Dima-Bulavenko** [comment]
> > It seems the pattern already handles cases where `max_digits` or `decimal_places` is `None`.
> 
> Sorry what I said was confusing, I wanted to say that it would be nice to have simpler patterns generated for cases where `max_digits` xor `decimal_places` is `None`.

Thank you for clarifying your request.

I've updated the implementation to generate simpler patterns for each combination of **max_digits** and **decimal_places**.

File: pydantic/json_schema.py [hunk 3]
```diff
@@ -674,7 +674,49 @@ def decimal_schema(self, schema: core_schema.DecimalSchema) -> JsonSchemaValue:
...
+                    rf')'
+                )
+
+            # Case 3: Only decimal_places is set
+            elif max_digits is None and decimal_places is not None:
+                pattern += rf'\d*\.?\d{{0,{decimal_places}}}0*$'
+
+            # Case 4: Both are None (no restrictions)
+            else:
+                pattern += r'\d*\.?\d*$'  # look for arbitrary integer or decimal
```
Comment (Viicos) [inline pydantic/json_schema.py:715]: This allows `'.'` as an input, which can't be validated.

File: pydantic/json_schema.py [hunk 4]
```diff
@@ -674,7 +674,49 @@ def decimal_schema(self, schema: core_schema.DecimalSchema) -> JsonSchemaValue:
...
+                    rf'(?:'
+                    rf'\d{{0,{integer_places}}}'
+                    rf'|'
+                    rf'(?=[\d.]{{1,{max_digits + 1}}}0*$)'
+                    rf'\d{{0,{integer_places}}}\.\d{{0,{decimal_places}}}0*$'
+                    rf')'
+                )
+
+            # Case 2: Only max_digits is set
+            elif max_digits is not None and decimal_places is None:
```
Comment (Viicos) [inline pydantic/json_schema.py:699]: With `max_digits` set to e.g. 5, it wrongfully matches `'1000.1111111'`, etc.

**Review by Viicos — COMMENTED** [review]
Given the complexity to implement this correctly (see my comments above), I'm starting to think if we should just define a simpler regex, describing the allowed characters (something like `^(?:\+|-)?[\d.]+$` in validation mode, and `^-?[\d.]+$` in serialization mode). This regex will also allow invalid inputs (but at least we are certain valid inputs wont be disallowed).

File: pydantic/json_schema.py (hunk 3 above)
Comment (Dima-Bulavenko) [reply pydantic/json_schema.py:715]: Thank you for pointing this out.
However, `r'\d*\.?\d*$'` is not the full pattern used for validation.

The complete pattern includes an additional regex component that prevents `'.'` from being a valid input:
https://github.com/pydantic/pydantic/blob/88adb2d4806a37fd21ae06f88d4670190759974f/pydantic/json_schema.py#L682-L684

Additionally, the edge case for `'.'` is covered by the following test:
https://github.com/pydantic/pydantic/blob/88adb2d4806a37fd21ae06f88d4670190759974f/tests/test_json_schema.py#L7005-L7009

File: pydantic/json_schema.py (hunk 4 above)
Comment (Dima-Bulavenko) [reply pydantic/json_schema.py:699]: Thank you for your feedback.

From my testing, the pattern appears to work correctly when `max_digits=5` and the invalid value `'1000.1111111'` is rejected as expected.

Here is a test demonstrating this behavior:
```python
@pytest.fixture
def get_decimal_pattern():
    def pattern(max_digits=None, decimal_places=None) -> str:
        field = TypeAdapter(Annotated[Decimal, Field(max_digits=max_digits, decimal_places=decimal_places)])
        return field.json_schema()['anyOf'][1]['pattern']
    return pattern

@pytest.mark.parametrize('invalid_decimal', ['1000.1111111'])
def test_only_max_digits_set(invalid_decimal, get_decimal_pattern):
    pattern = get_decimal_pattern(max_digits=5, decimal_places=None)
    assert re.fullmatch(pattern, invalid_decimal) is None
```

Let me know if there's a specific case I may have missed!

**Comment by Dima-Bulavenko** [comment]
I think in validation mode, it's better to use the pattern `^[-+]?(?:\d+|\d+\.\d+)$`, which allows both integers and decimals of any size. However, if we use this simple pattern, we completely lose the main purpose of this issue (#10867), which is to check for `max_digits` and `decimal_places`.

Also, the more complex regex patterns are only created in two cases:
1. When `max_digits` is set, or
2. When both `max_digits` and `decimal_places` are not `None`.

I believe these cases don't come up very often, but if a user sets these options, they probably want the regex pattern to actually check `max_digits` and `decimal_places`.

For serialization mode, I agree that a simple regex like `^-?(?:\d+|\d+\.\d+)$` is useful because it shows that the data can be either an integer or a decimal.

Please let me know what you think about this. I should also mention that it's possible I don't fully understand the purpose of the validation and serialization schemas or all the cases where they might be used.

**Review by Viicos — APPROVED** [review]
Thanks @Dima-Bulavenko, @Jack70248 I'm adding you as a co-author for the initial work.
